File: menubar/BrewBar/Sources/Services/CrashReporter.swift
import Foundation
import Security
import os

private let crashLogger = Logger(subsystem: "com.molinesdesigns.brewbar", category: "CrashReporter")

// File-private free function so it can be passed as a C function pointer
// to NSSetUncaughtExceptionHandler (no captured context allowed).
private let handleUncaught: @convention(c) (NSException) -> Void = { exception in
    CrashReporter.didReceiveUncaughtException(exception)
}

/// Opt-in crash reporting that posts JSON to a user-configured endpoint
/// (typically a self-hosted dashboard on the user's NAS).
///
/// Configuration is read from UserDefaults except for the bearer token, which
/// lives in Keychain (BK-011). Endpoint and enabled flag are not secrets:
///   defaults write com.molinesdesigns.brewbar crashReporterEndpoint https://nas.local:port/crash
///   defaults write com.molinesdesigns.brewbar crashReporterEnabled -bool YES
/// To set the token from Settings.app or a script, see `setToken(_:)`.
///
/// Note: this captures Objective-C `NSException` and explicitly reported
/// errors. Pure-Swift `fatalError` / runtime traps cannot be intercepted
/// without a native crash SDK.
// QA-007: seam for plugging in a real crash SDK (Sentry, Crashlytics, Bugsnag).
// Pure-Swift `fatalError`/runtime traps cannot be intercepted from inside the
// process — only a native handler installed via Sentry-Cocoa or equivalent can
// catch them. Until that PR lands, anyone wanting third-party crash capture
// implements this protocol and registers it via `CrashReporter.installSDK(_:)`.
//
// Keeping this protocol in-tree avoids tying the codebase to a specific SDK
// while letting the call sites (`CrashReporter.report(_:)`) forward to it.
protocol CrashReportingSDK: Sendable {
    func captureException(_ exception: NSException, context: [String: String])
    func captureError(_ error: Error, context: [String: String])
}

enum CrashReporter {
    private static let endpointKey = "crashReporterEndpoint"
    private static let enabledKey = "crashReporterEnabled"
    private static let keychainService = "com.molinesdesigns.brewbar.crashReporter"
    private static let keychainAccount = "token"
    private static let postTimeout: TimeInterval = 5

    nonisolated(unsafe) private static var externalSDK: (any CrashReportingSDK)?

    /// Register a third-party SDK. Pass nil to detach. See QA-007.
    static func installSDK(_ sdk: (any CrashReportingSDK)?) {
        externalSDK = sdk
    }

    private struct Report: Encodable {
        let app: String
        let version: String
        let platform: String
        let os: String
        let arch: String
        let machineId: String
        let timestamp: String
        let level: String
        let message: String
        let stack: String?
        let context: [String: String]
    }

    private static var endpoint: URL? {
        guard let raw = UserDefaults.standard.string(forKey: endpointKey)?
            .trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty
        else { return nil }
        return URL(string: raw)
    }

    private static var token: String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: keychainAccount,
            kSecMatchLimit as String: kSecMatchLimitOne,
            kSecReturnData as String: true,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess,
              let data = item as? Data,
              let value = String(data: data, encoding: .utf8),
              !value.isEmpty
        else { return nil }
        return value
    }

    /// Persist the bearer token in Keychain. Pass nil to clear.
    static func setToken(_ token: String?) {
        let baseQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: keychainAccount,
        ]
        // Always delete first so updates do not collide with stale entries.
        SecItemDelete(baseQuery as CFDictionary)
        guard let token, !token.isEmpty, let data = token.data(using: .utf8) else { return }
        var insert = baseQuery
        insert[kSecValueData as String] = data
        insert[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let status = SecItemAdd(insert as CFDictionary, nil)
        if status != errSecSuccess {
            crashLogger.error("Could not store crash-reporter token: \(status, privacy: .public)")
        }
    }

    private static var isEnabled: Bool {
        UserDefaults.standard.bool(forKey: enabledKey) && endpoint != nil
    }

    /// Install global handlers. Safe to call from `applicationDidFinishLaunching`.
    static func install() {
        // QA-007: even when the NAS endpoint is disabled, we still want the
        // SDK forward to fire if one is registered. Both paths are independent.
        let sdkActive = externalSDK != nil
        guard isEnabled || sdkActive else { return }
        // The C function pointer NSSetUncaughtExceptionHandler expects cannot
        // capture context, so the handler reads the static SDK + posts the
        // NAS report through the static `send` entrypoint.
        NSSetUncaughtExceptionHandler(handleUncaught)
        crashLogger.info("CrashReporter enabled (NAS: \(isEnabled), SDK: \(sdkActive))")
    }

    /// Internal callback for the C uncaught-exception handler.
    static func didReceiveUncaughtException(_ exception: NSException) {
        externalSDK?.captureException(exception, context: ["kind": "NSException"])
        send(
            level: "fatal",
            message: exception.reason ?? exception.name.rawValue,
            stack: exception.callStackSymbols.joined(separator: "\n"),
            context: ["kind": "NSException"]
        )
    }

    /// Manually report a non-fatal error.
    static func report(_ error: Error, context: [String: String] = [:]) {
        externalSDK?.captureError(error, context: context)
        guard isEnabled else { return }
        send(
            level: "error",
            message: error.localizedDescription,
            stack: Thread.callStackSymbols.joined(separator: "\n"),
            context: context
        )
    }

    // MARK: - Internal

    private static func send(level: String, message: String, stack: String?, context: [String: String]) {
        guard let endpoint, isAllowedHost(endpoint) else { return }

        let report = Report(
            app: "brewbar",
            version: marketingVersion(),
            platform: "macos",
            os: ProcessInfo.processInfo.operatingSystemVersionString,
            arch: hardwareArch(),
            machineId: machineId(),
            timestamp: ISO8601DateFormatter().string(from: Date()),
            level: level,
            message: message,
            stack: stack,
            context: context
        )

        guard let body = try? JSONEncoder().encode(report) else { return }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.timeoutInterval = postTimeout
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        request.httpBody = body

        // Fire-and-forget. Synchronous wait would block the host process.
        URLSession.shared.dataTask(with: request) { _, _, error in
            if let error {
                crashLogger.warning("crash POST failed: \(error.localizedDescription, privacy: .public)")
            }
        }.resume()
    }

    private static func isAllowedHost(_ url: URL) -> Bool {
        if url.scheme == "https" { return true }
        if url.scheme == "http", let host = url.host {
            // Allow plain http only for loopback / private LAN destinations.
            return host == "localhost"
                || host.hasPrefix("127.")
                || host.hasPrefix("10.")
                || host.hasPrefix("192.168.")
                || host.hasPrefix("172.16.") || host.hasPrefix("172.17.")
                || host.hasPrefix("172.18.") || host.hasPrefix("172.19.")
                || host.hasPrefix("172.20.") || host.hasPrefix("172.21.")
                || host.hasPrefix("172.22.") || host.hasPrefix("172.23.")
                || host.hasPrefix("172.24.") || host.hasPrefix("172.25.")
                || host.hasPrefix("172.26.") || host.hasPrefix("172.27.")
                || host.hasPrefix("172.28.") || host.hasPrefix("172.29.")
                || host.hasPrefix("172.30.") || host.hasPrefix("172.31.")
                || host == "::1"
        }
        return false
    }

    private static func machineId() -> String {
        let path = NSHomeDirectory() + "/.brew-tui/machine-id"
        if let raw = try? String(contentsOfFile: path, encoding: .utf8) {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if !trimmed.isEmpty { return trimmed }
        }
        return "unknown"
    }

    private static func marketingVersion() -> String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "unknown"
    }

    private static func hardwareArch() -> String {
        var size: Int = 0
        sysctlbyname("hw.machine", nil, &size, nil, 0)
        var machine = [CChar](repeating: 0, count: size)
        sysctlbyname("hw.machine", &machine, &size, nil, 0)
        return String(cString: machine)
    }
}

File: src/lib/crash-reporter.ts
import { readFile } from 'node:fs/promises';
import { homedir, platform, release, arch } from 'node:os';
import { join } from 'node:path';
import { logger } from '../utils/logger.js';
import { getMachineId } from './data-dir.js';

// Opt-in crash reporting. The TUI never sends data unless the user has
// configured an endpoint, either via the env var below or via the persisted
// config file at ~/.brew-tui/crash-reporter.json. The dashboard backend is
// expected to live on the user's own infrastructure (NAS).
const ENDPOINT_ENV = 'BREW_TUI_CRASH_ENDPOINT';
const TOKEN_ENV = 'BREW_TUI_CRASH_TOKEN';
const CONFIG_PATH = join(homedir(), '.brew-tui', 'crash-reporter.json');
const POST_TIMEOUT_MS = 5_000;

interface CrashReporterConfig {
  endpoint: string | null;
  token: string | null;
  enabled: boolean;
}

interface CrashReport {
  app: 'brew-tui';
  version: string;
  platform: string;
  os: string;
  arch: string;
  machineId: string;
  timestamp: string;
  level: 'fatal' | 'error';
  message: string;
  stack: string | null;
  context: Record<string, unknown>;
}

let _installed = false;

async function loadConfigFromDisk(): Promise<CrashReporterConfig> {
  try {
    const raw = await readFile(CONFIG_PATH, 'utf-8');
    const parsed = JSON.parse(raw) as Partial<CrashReporterConfig>;
    return {
      endpoint: typeof parsed.endpoint === 'string' ? parsed.endpoint : null,
      token: typeof parsed.token === 'string' ? parsed.token : null,
      enabled: parsed.enabled === true,
    };
  } catch {
    return { endpoint: null, token: null, enabled: false };
  }
}

async function resolveConfig(): Promise<CrashReporterConfig> {
  const envEndpoint = process.env[ENDPOINT_ENV]?.trim();
  const envToken = process.env[TOKEN_ENV]?.trim();
  if (envEndpoint) {
    return { endpoint: envEndpoint, token: envToken ?? null, enabled: true };
  }
  return loadConfigFromDisk();
}

function isHttpsOrLocal(url: string): boolean {
  try {
    const parsed = new URL(url);
    if (parsed.protocol === 'https:') return true;
    // Allow plain http only for loopback/LAN destinations (NAS).
    if (parsed.protocol === 'http:') {
      const host = parsed.hostname;
      return /^(localhost|127\.\d+\.\d+\.\d+|10\.|192\.168\.|172\.(1[6-9]|2\d|3[0-1])\.|::1$)/i.test(host);
    }
    return false;
  } catch {
    return false;
  }
}

async function postReport(report: CrashReport, config: CrashReporterConfig): Promise<void> {
  if (!config.endpoint || !isHttpsOrLocal(config.endpoint)) return;
  try {
    await fetch(config.endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...(config.token ? { 'Authorization': `Bearer ${config.token}` } : {}),
      },
      body: JSON.stringify(report),
      signal: AbortSignal.timeout(POST_TIMEOUT_MS),
    });
  } catch (err) {
    // Never let the reporter itself crash the host process.
    logger.warn('crash-reporter: POST failed', { error: String(err) });
  }
}

function buildReport(level: 'fatal' | 'error', err: unknown, context: Record<string, unknown>, machineId: string, version: string): CrashReport {
  const message = err instanceof Error ? err.message : String(err);
  const stack = err instanceof Error && typeof err.stack === 'string' ? err.stack : null;
  return {
    app: 'brew-tui',
    version,
    platform: platform(),
    os: release(),
    arch: arch(),
    machineId,
    timestamp: new Date().toISOString(),
    level,
    message,
    stack,
    context,
  };
}

/**
 * Manually report a non-fatal error. No-op if reporting is not enabled.
 */
export async function reportError(err: unknown, context: Record<string, unknown> = {}): Promise<void> {
  const config = await resolveConfig();
  if (!config.enabled || !config.endpoint) return;
  const machineId = await getMachineId();
  const version = typeof process.env.APP_VERSION === 'string' ? process.env.APP_VERSION : 'unknown';
  await postReport(buildReport('error', err, context, machineId, version), config);
}

/**
 * Install global handlers for uncaughtException / unhandledRejection.
 * No-op when the reporter is not enabled. Safe to call multiple times.
 */
export async function installCrashReporter(): Promise<void> {
  if (_installed) return;
  const config = await resolveConfig();
  if (!config.enabled || !config.endpoint) return;
  _installed = true;

  const machineId = await getMachineId();
  const version = typeof process.env.APP_VERSION === 'string' ? process.env.APP_VERSION : 'unknown';

  process.on('uncaughtException', (err) => {
    void postReport(buildReport('fatal', err, { kind: 'uncaughtException' }, machineId, version), config);
  });

  process.on('unhandledRejection', (reason) => {
    void postReport(buildReport('error', reason, { kind: 'unhandledRejection' }, machineId, version), config);
  });

  logger.info('crash-reporter: enabled', { endpoint: config.endpoint });
}
